core/modules/dashboard/dashboard.module

```diff
@@ -0,0 +1,444 @@
+    'info' => t('Dashboard - Content types'),
+    'description' => t('Provides links to manage content types'),
+    'class' => 'DashboardContentTypesBlock',
+  );
+  $blocks['overview_content'] = array(
+    'info' => t('Dashboard - Content'),
+    'description' => t('Displays a summary of content statistics'),
+    'class' => 'DashboardOverviewContentBlock',
+  );
+  $blocks['overview_user'] = array(
+    'info' => t('Dashboard - User accounts'),
+    'description' => t('Displays a summary of user account statistics'),
+    'class' => 'DashboardOverviewUserBlock',
+  );
+  $blocks['updates'] = array(
+    'info' => t('Dashboard - Available updates'),
+    'description' => t('Displays a list of available updates for modules, themes, layouts, and Backdrop core.'),
+    'class' => 'DashboardUpdateBlock',
+  );
+
+  if (module_exists('taxonomy')) {
+    $blocks['taxonomy'] = array(
+      'info' => t('Dashboard - Categories'),
+      'description' => t('Provides links to manage taxonomy'),
+      'class' => 'DashboardTaxonomyBlock',
+    );
+  }
+
+  return $blocks;
+}
+
+/**
+ * Implements hook_autoload_info().
+ */
+function dashboard_autoload_info() {
+  return array(
+    'DashboardCreateBlock' => 'includes/block.create.inc',
+    'DashboardMenuBlock' => 'includes/block.menus.inc',
+    'DashboardContentTypesBlock' => 'includes/block.node_types.inc',
+    'DashboardOverviewContentBlock' => 'includes/block.overview_content.inc',
+    'DashboardOverviewUserBlock' => 'includes/block.overview_user.inc',
+    'DashboardTaxonomyBlock' => 'includes/block.taxonomy.inc',
+    'DashboardUpdateBlock' => 'includes/block.update.inc',
+  );
+}
+
+/**
+ * Implements hook_user_presave().
+ */
+function dashboard_user_presave($account) {
+  cache_clear_all('dashboard_user_overview', 'cache');
+}
+
+/**
+ * Implements hook_user_delete().
+ */
+function dashboard_user_delete($account) {
+  cache_clear_all('dashboard_user_overview', 'cache');
+}
+
+/**
+ * Implements hook_user_login().
+ */
+function dashboard_user_login(&$edit, $account) {
+  // Don't redirect if a destination is already set.
+  if (!empty($_GET['destination'])) {
+    return;
+  }
+
+  // Don't redirect for password reset forms.
+  if (isset($_POST['form_id']) && $_POST['form_id'] == 'user_pass_reset') {
+    return;
+  }
+
+  // Don't redirect if no permission to access dashboard.
+  if (!user_access('access dashboard', $account)) {
+    return;
+  }
+
+  // Don't redirect if redirect setting is not set.
+  if (config_get('dashboard.settings', 'dashboard_login_redirect') != 1) {
+    return;
+  }
+
+  // Check access to layout and redirect.
+  $dashboard = layout_load('dashboard');
+  if ($dashboard && !$dashboard->disabled && $dashboard->checkAccess()) {
+    $_GET['destination'] = 'admin/dashboard';
+  }
+}
+
+/**
+ * Implements hook_preprocess_page().
+ */
+function dashboard_preprocess_page(&$variables) {
+  if (arg(0) == 'admin' && arg(1) == 'dashboard') {
```

Comment:
Seems that we could/should use `preprocess_layout` instead of `preprocess_page` so we don't have to do URL checking.

Reply:
yep. fixed by @quicksketch